pkg/controller/elasticsearch/filesettings/reconciler.go

```diff
@@ -51,27 +54,33 @@ func ReconcileEmptyFileSettingsSecret(
 		return err
 	}
 
-	return ReconcileSecret(ctx, c, expectedSecret, es)
+	return ReconcileSecret(ctx, c, expectedSecret, &es)
 }
 
-// ReconcileSecret reconciles the given file settings Secret for the given Elasticsearch.
+// ReconcileSecret reconciles the given Secret.
 // This implementation is slightly different from reconciler.ReconcileSecret to allow resetting managed annotations.
 func ReconcileSecret(
 	ctx context.Context,
 	c k8s.Client,
 	expected corev1.Secret,
-	es esv1.Elasticsearch,
+	owner client.Object,
 ) error {
 	reconciled := &corev1.Secret{}
 	return reconciler.ReconcileResource(reconciler.Params{
 		Context:    ctx,
 		Client:     c,
-		Owner:      &es,
+		Owner:      owner,
 		Expected:   &expected,
 		Reconciled: reconciled,
 		NeedsUpdate: func() bool {
-			return !maps.IsSubset(expected.Labels, reconciled.Labels) ||
-				!maps.IsSubset(expected.Annotations, reconciled.Annotations) ||
+			/*
+				log := ulog.FromContext(ctx)
+				log.Info("Reconciled is", "reconciled", *reconciled)
+				log.Info("Expected is", "expected", expected)
+			*/
+			// needs update if expected labels/annotations are not a subset of the existing labels/annotions or if managed labels/annotations have been removed/changed or if the data itself has changed.
+			return (!maps.IsSubset(expected.Labels, reconciled.Labels) || !maps.IsEqualSubset(expected.Labels, reconciled.Labels, managedLabels)) ||
+				(!maps.IsSubset(expected.Annotations, reconciled.Annotations) || !maps.IsEqualSubset(expected.Annotations, reconciled.Annotations, managedAnnotations)) ||
 				!reflect.DeepEqual(expected.Data, reconciled.Data)
```

Comment:
Not from your PR but I just realized that this comparison is not effective when keys are empty, because `DeepEquals` does make the difference between an empty slice and a `nil` slice:

![image](https://github.com/elastic/cloud-on-k8s/assets/976373/94226a30-b5ee-407c-98d4-d87a1c714f29)

⬆️ This continuously returns `true`

 Or another way of looking at things would be to consider that we should not try to set unnecessary keys in a Secret as it is done in [newElasticsearchConfigSecret](https://github.com/elastic/cloud-on-k8s/blob/45782fb3fdd06e95fbc30e0b5efa31233f798c52/pkg/controller/stackconfigpolicy/elasticsearch_config_settings.go#L39):

Reply:
I made changes to add keys only when data is present.